test/integration/consul-container/test/envoy_extensions/wasm_test.go

```diff
@@ -0,0 +1,627 @@
+	defaults := api.ServiceConfigEntry{
+		Kind:     api.ServiceDefaults,
+		Name:     "static-server",
+		Protocol: "http",
+		EnvoyExtensions: []api.EnvoyExtension{{
+			Name: "builtin/wasm",
+			Arguments: map[string]any{
+				"Protocol":     "http",
+				"ListenerType": "inbound",
+				"PluginConfig": map[string]any{
+					"VmConfig": map[string]any{
+						"Code": map[string]any{
+							"Local": map[string]any{
+								"Filename": "/wasm_add_header.wasm",
+							},
+						},
+					},
+				},
+			},
+		}},
+	}
+
+	_, _, err = consul.ConfigEntries().Set(&defaults, nil)
+	require.NoError(t, err, "could not set config entries")
+
+	// Check that header is present after wasm applied
+	c2 := cleanhttp.DefaultClient()
+
+	// The wasm plugin is not always applied on the first call. Retry and see if it is loaded.
+	retryStrategy := func() *retry.Timer {
+		return &retry.Timer{Timeout: 5 * time.Second, Wait: time.Second}
+	}
+	retry.RunWith(retryStrategy(), t, func(r *retry.R) {
+		res2, err := c2.Get(fmt.Sprintf("http://localhost:%d", port))
+		require.NoError(r, err)
+
+		if value := res2.Header.Get(http.CanonicalHeaderKey("x-test")); value == "" {
+			r.Fatal("test header missing after WASM applied")
+		}
+	})
+}
+
+func createServices(t *testing.T, cluster *libcluster.Cluster) (libservice.Service, libservice.Service) {
+	node := cluster.Agents[0]
+	client := node.GetClient()
+	// Create a service and proxy instance
+	serviceOpts := &libservice.ServiceOpts{
+		Name:     libservice.StaticServerServiceName,
+		ID:       libservice.StaticServerServiceName,
+		HTTPPort: 8080,
+		GRPCPort: 8079,
+	}
+	cwd, err := os.Getwd()
+	require.NoError(t, err, "could not get current working directory")
+	hostWASMDir := fmt.Sprintf("%s/testdata/wasm_test_files", cwd)
+
+	wasmFile := testcontainers.ContainerFile{
+		HostFilePath:      fmt.Sprintf("%s/wasm_add_header.go.wasm", hostWASMDir),
+		ContainerFilePath: "/wasm_add_header.wasm",
+		FileMode:          777,
+	}
+
+	customFn := chain(
+		copyFilesToContainer([]testcontainers.ContainerFile{wasmFile}),
+		chownFiles([]testcontainers.ContainerFile{wasmFile}, "envoy", true),
+	)
+
+	// Create a service and proxy instance
+	_, staticProxy, err := libservice.CreateAndRegisterStaticServerAndSidecarWithCustomContainerConfig(node, serviceOpts, customFn)
+	require.NoError(t, err)
+
+	libassert.CatalogServiceExists(t, client, "static-server-sidecar-proxy", nil)
+	libassert.CatalogServiceExists(t, client, libservice.StaticServerServiceName, nil)
+
+	// Create a client proxy instance with the server as an upstream
+
+	clientConnectProxy, err := libservice.CreateAndRegisterStaticClientSidecar(node, "", false, false)
+	require.NoError(t, err)
+
+	libassert.CatalogServiceExists(t, client, "static-client-sidecar-proxy", nil)
+
+	return clientConnectProxy, staticProxy
+}
+
+// buildWASM uses a tinygo build container to compile wasm from source and then output a binary.
+func buildWASM(t *testing.T, hostWASMDir string) {
```

Comment:
Building image on-the-fly might cause some flakiness in the integration test since sometimes it takes a while to download the packages and build the binary. We had experienced this issue before, so we extracted the step outside of the test code and moved it into a dedicated the step https://github.com/hashicorp/consul/blob/94eb36b7750d1156d4203e34fb1c85872a3c8e6d/.github/workflows/test-integrations.yml#L510-L513

Maybe we can apply the same methodology in this test.

Reply:
Makes sense. I think I would like to keep it out of github actions specifically and maybe move it into make. Ideally, someone should be able to edit the example wasm if needed then run their `make` or test command without having to worry about the updates to the wasm binary. 

How would you feel about using a make task to build the binary which the integration test task depends on?

Reply:
> How would you feel about using a make task to build the binary which the integration test task depends on?

sounds good to me assuming that we will use the `make wasm-image` as one step in the job.

Reply:
Moved the wasm compilation into its own step and am building it directly rather than in a docker build container. Was having some weird issues with tinygo panicking in the docker container on the runner. This seems to work better. 

https://github.com/hashicorp/consul/pull/17756/files#diff-d678eaee969dd3d0e446036483bd39ddf3017abad369cdc63ee0ff0b17ed455aR308

Reply:
@huikang Please take another look. We were seeing intermittent failures during compilation on the build runners. Some runners would panic while compiling with tinygo while others were fine. We decided to move forward with the simple path and add the compiled wasm to the test data and include instructions on how to rebuild as needed.